File: java/src/main/java/fr/sdecout/handson/rest/library/PostalCode.java
package fr.sdecout.handson.rest.library;

import jakarta.annotation.Nonnull;

import java.util.regex.Pattern;

import static java.util.Objects.requireNonNull;

public record PostalCode(@Nonnull String value) {

    private static final Pattern PATTERN = Pattern.compile("[0-9]{5}");

    public PostalCode {
        requireNonNull(value);
        if (!PATTERN.matcher(value).matches()) { throw new IllegalArgumentException("Postal code must contain 5 numeric digits"); };
    }

    public String departmentCode() {
        return value.substring(0, 2);
    }

    @Override
    public String toString() {
        return value;
    }
}

File: kotlin/src/main/kotlin/fr/sdecout/handson/rest/library/PostalCode.kt
package fr.sdecout.handson.rest.library

@JvmInline
value class PostalCode(val value: String) {

    init {
        require(value.matches(Regex("[0-9]{5}"))) { "Postal code must contain 5 numeric digits" }
    }

    val departmentCode get() = value.substring(0..1)

    override fun toString() = value

}
